connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkResult.scala

```diff
@@ -142,24 +150,41 @@ private[sql] class SparkResult[T](
   /**
    * Returns an iterator over the contents of the result.
    */
-  def iterator: java.util.Iterator[T] with AutoCloseable = {
+  def iterator: java.util.Iterator[T] with AutoCloseable =
+    buildIterator(destructive = false)
+
+  /**
+   * Returns an destructive iterator over the contents of the result.
+   */
+  def destructiveIterator: java.util.Iterator[T] with AutoCloseable =
+    buildIterator(destructive = true)
+
+  private def buildIterator(destructive: Boolean): java.util.Iterator[T] with AutoCloseable = {
     new java.util.Iterator[T] with AutoCloseable {
       private[this] var batchIndex: Int = -1
       private[this] var iterator: java.util.Iterator[InternalRow] = Collections.emptyIterator()
       private[this] var deserializer: Deserializer[T] = _
+
       override def hasNext: Boolean = {
         if (iterator.hasNext) {
           return true
         }
+
         val nextBatchIndex = batchIndex + 1
-        val hasNextBatch = if (nextBatchIndex == batches.size) {
+        if (destructive && idxToBatches.contains(batchIndex)) {
+          val currentBatch = idxToBatches(batchIndex)
```

Comment:
if `currentBatch` just used to close `ColumnarBatch`, how about
```
        if (destructive) {
          idxToBatches.remove(batchIndex).foreach(_.close())
        }
```

Reply:
Yes, that looks better. Thanks.